src/ripple/app/misc/impl/ValidatorList.cpp

```diff
@@ -17,117 +17,611 @@
-    std::lock_guard <std::mutex> sl (mutex_);
+    if (fetchedLists_ == 0)
+        return 0;
 
-    if (ephemeral_.find (identity) != ephemeral_.end())
+    // If only one list is available, use all available validators in that list
+    if (fetchedLists_ == 1)
     {
-        JLOG (j_.error()) <<
-            toBase58 (TokenType::TOKEN_NODE_PUBLIC, identity) <<
-            ": permanent key exists in ephemeral table!";
-        return false;
+        for (auto const& list : publisher_lists_)
+        {
+            if (list.second.list.size())
+                return list.second.list.size();
+        }
     }
 
-    return permanent_.emplace (identity, comment).second;
+    // Target trusted validator size is
+    // 80% of the median recommended validator list size
+    std::vector<int> listSizes;
+    listSizes.reserve (publisher_lists_.size ());
+    for (auto const& pubList : publisher_lists_)
+    {
+        if (! pubList.second.list.empty())
+            listSizes.push_back (pubList.second.list.size());
+    }
+
+    std::sort (listSizes.begin(), listSizes.end());
+
+    auto iter = std::next(listSizes.begin (), listSizes.size()/2);
+    auto medianSize = *iter;
+    if ((listSizes.size () % 2) == 0)
+        medianSize = (medianSize + (*--iter)) / 2;
+
+    return medianSize * 0.8;
 }
 
-bool
-ValidatorList::removePermanentKey (
-    PublicKey const& identity)
+int
+ValidatorList::calcQuorum (
+    std::uint32_t const& nTrusted,
+    std::uint32_t const& nListed)
 {
-    std::lock_guard <std::mutex> sl (mutex_);
-    return permanent_.erase (identity);
+    const std::vector<int> neededValidations{
+        0, 1, 2, 3, 3, 4, 5, 5, 6, 7, 7};
+    auto constexpr quorumRatio = 0.8;
+
+    // Floor of 80% of minimum required overlap with all listed validators
+    auto quorum = std::ceil (
+        quorumRatio *
+        std::ceil (
+            MIN_NETWORK_VALIDATOR_LIST_OVERLAP *
+            std::max (nListed, nTrusted)));
+
+    // If >10 trusted validators, set quorum to 80%
+    // of validations from previous ledger.
+    // Otherwise, use lookup table with quorums for
+    // byzantine fault tolerance (3f+1)
+    int nVals = (nTrusted > 10) ?
+        std::ceil (nTrusted * quorumRatio) :
+        neededValidations[nTrusted];
+
+    return (quorum < nVals) ? nVals : quorum;
+}
+
+// should this not count our own validator key?
+void
+ValidatorList::update (
+    ValidationSet const& activeValidators)
+{
+    std::lock_guard <std::mutex> lock{mutex_};
+
+    std::multimap<int, PublicKey> rankedValidators;
+
+    // "Iterate" the listed validators in random order so that it is not
+    // deterministic which validators tied with the same number of listings
+    // make the cut to be included on the trusted list
+    std::vector<int> indexes;
+    indexes.reserve (validatorListings_.size());
+    for (int i=0; i<validatorListings_.size(); ++i)
+        indexes.push_back (i);
+
+    std::random_shuffle (indexes.begin(), indexes.end());
```

Comment:
This should be:

    std::shuffle (indexes.begin(), index.end(), crypto_prng());